### PR TITLE
fix locks to use atomic locks

### PR DESCRIPTION
**Pull request by patrickweh**

### CODE REVIEW

File: src/Helpers/LockHelper.php
Comment (SirSplasch) [inline src/Helpers/LockHelper.php:1]: not quite sure if this helper is really needed

File: src/Actions/Locking/UnlockModel.php
Comment (SirSplasch) [inline src/Actions/Locking/UnlockModel.php:1]: Please keep this action

File: src/Http/Controllers/LockController.php
Comment (SirSplasch) [inline src/Http/Controllers/LockController.php:1]: re-use actions in lock and unlock functions

File: src/Traits/Lockable.php [hunk 1]
```diff
@@ -3,56 +3,131 @@
...
 
-    public function lock(): MorphOne
+    public function lock(): bool
     {
-        return $this->morphOne(Lock::class, 'model');
+
+        $lockDuration = config('session.lifetime') * 60;
+        $lock = Cache::lock($this->getLockName(), $lockDuration);
+
+        if ($isLocked = $lock->get()) {
```
Comment (SirSplasch) [inline src/Traits/Lockable.php:59]: decide whether it is isLocked or hasLock

File: tests/Feature/Api/LockTest.php [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        Sanctum::actingAs($this->user, ['user']);
+        $this->users[0]->lock();
+
+        $response = $this->actingAs($this->user)
+            ->post('/api/force-unlock', ['model_type' => User::class, 'model_id' => $this->users[0]->id]);
+
+        $response->assertStatus(204);
+    }
+
+    public function test_write_locked_record()
```
Comment (SirSplasch) [inline tests/Feature/Api/LockTest.php:93]: rename to test_update_locked_record

File: tests/Feature/Api/LockTest.php [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        $user->firstname = Str::random();
+
+
+        $response = $this->actingAs($this->user)
+            ->put('/api/users', $user->toArray());
+
+        $response->assertStatus(423);
+    }
+
+    public function test_write_own_locked_record()
```
Comment (SirSplasch) [inline tests/Feature/Api/LockTest.php:109]: rename to test_update_own_locked_record

File: tests/Feature/Api/LockTest.php [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        $this->user->givePermissionTo($this->permissions['update_user']);
+        $this->user->givePermissionTo($this->permissions['lock']);
+
+        Sanctum::actingAs($this->user, ['user']);
+
+        $response = $this->actingAs($this->user)
```
Comment (SirSplasch) [inline tests/Feature/Api/LockTest.php:116]: Test should be granular, not rely on other tests.

File: src/Models/User.php [hunk 5]
```diff
@@ -96,9 +98,9 @@ public function language(): BelongsTo
         return $this->belongsTo(Language::class);
     }
 
-    public function locks(): MorphMany
+    public function locks(): LockHelper
     {
-        return $this->morphMany(Lock::class, 'authenticatable');
+        return new LockHelper($this);
```
Comment (SirSplasch) [inline src/Models/User.php:103]: I don't think I'm a big fan of this.

File: src/Http/Controllers/LockController.php [hunk 6]
```diff
@@ -2,112 +2,59 @@
...
-            ->paginate($perPage);
-
-        return ResponseHelper::createResponseFromBase(statusCode: 200, data: $locks);
+        $model = $request->get('model_type')::query()
+            ->whereKey($request->get('model_id'))
+            ->firstOrFail();
+
+        return $model->lock() || auth()->user()->is($model->getLockedBy())
+            ? ResponseHelper::createResponseFromBase(
+                statusCode: 201,
```
Comment (SirSplasch) [inline src/Http/Controllers/LockController.php:30]: why change status code to 201? I prefer 200.

File: src/Http/Controllers/LockController.php [hunk 7]
```diff
@@ -2,112 +2,59 @@
...
+            : ResponseHelper::locked(
+                'record is already locked by ' . $model->getLockedBy()?->name ?? 'unknown'
+            );
     }
 
-    /**
-     * @throws \Illuminate\Validation\ValidationException
-     */
-    public function lock(string $modelType, Request $request): JsonResponse
+    public function unlock(LockRecordRequest $request): JsonResponse
```
Comment (SirSplasch) [inline src/Http/Controllers/LockController.php:39]: why change status codes?

File: src/Actions/Lock/ForceUnlock.php [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+            ->filter(fn ($model) => in_array(Lockable::class, $model->traits->toArray()))
+            ->map(fn ($model) => $model->class)
+            ->toArray();
+    }
+
+    public function performAction(): mixed
+    {
+        $model = $this->data['model_type']::query()
+            ->whereKey($this->data['model_id'])
+            ->firstOrFail();
```
[the remark this answers is not in the document]
Comment (patrickweh) [reply src/Actions/Lock/ForceUnlock.php:31]: what happens when the model to unlock doesnt exist?

File: src/Actions/Lock/ForceUnlock.php [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+{
+    protected function boot(array $data): void
+    {
+        parent::boot($data);
+        $this->rules = (new LockRecordRequest())->rules();
+    }
+
+    public static function models(): array
+    {
+        return ModelInfo::forAllModels()
```
[the remark this answers is not in the document]
Comment (patrickweh) [reply src/Actions/Lock/ForceUnlock.php:20]: not sure, didnt have any problems while testing

File: tests/Feature/Api/LockTest.php (hunk 4 above)
Comment (patrickweh) [reply tests/Feature/Api/LockTest.php:116]: i dont know how to lock a record in behalf of an auth user without going through the api call

File: tests/Feature/Api/LockTest.php (hunk 3 above)
Comment (patrickweh) [reply tests/Feature/Api/LockTest.php:109]: √

**Pull request closed without merging**